File: mqp_colors.m
clear all; close all;
oimg = imread('35_1048.jpg');
[newimg, map] = rgb2ind(oimg, 16);
colormap(map)
hsvmap = rgb2hsv(map)

[rowCount,colcount,junk] = size(hsvmap);
h = hsvmap(:,1);
s = hsvmap(:,2);
v = hsvmap(:,3);


edges = [0, 0.053, 0.1083, 0.1805, 0.4444, 0.5277, 0.7138, 0.8055, 0.8888, 1]
colors = discretize(h,edges, 'categorical', {'red' 'orange' 'yellow' 'green' 'cyan' 'blue' 'purple' 'magenta' 'red'});

%colors = zeros(1, length(hsvmap));
%colors(or(h < 0.25, h > 0.95)) = 'red';
%colors(h >= 0.25 & h < 0.35) = 'orange';
%colors(h >= 0.25 && h < 0.95) = 'red';
%colors(h >= 0.25 && h < 0.95) = 'red';

colors = addcats(colors, {'black' 'white' 'brown' 'grey' }, 'After', 'orange');

colors(v < 0.25) = 'black';
colors(s < 0.05 & v > 0.95) = 'white';
colors((h >= 0.53 & h < 0.1083) & (v > .25 & v < .65)) = 'brown';
colors(isundefined(colors)) = 'grey';

finalRow = 160;
finalColumn = 160;
[nrows,ncols,junk] = size(newimg)
chunkRow = ceil(nrows / finalRow);
chunkColumn = ceil(ncols / finalColumn);
resimg = mat2tiles(newimg, chunkRow, chunkColumn);

resimg2 = cellfun(@(x) mode(x, 'all'), resimg,'UniformOutput',false);

finalcatimg = categorical(cell2mat(resimg2),[0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15], cellstr(colors));
finalnumimg = categorical(finalcatimg,{'red' 'orange' 'yellow' 'green' 'cyan' 'blue' 'purple' 'magenta' 'black' 'white' 'brown' 'grey'},{'1' '2' '3' '4' '5' '6' '7' '8' '9' '10' '11' '12'});
%now take those averaged colors and then divide them up into sections and find the most dominant color.
%B = reshape(newimg,25, [])
%resimg = imresize(newimg, [320 NaN]);
%I want to create the chunks based on the number of pixels the final output
%is set to be. So I need a variable for finalRow, finalColumn. I need to
%divide newimg rows and columns by finalRow and finalColumn to get how big
%each chunk is supposed to be, which means I need variables chunkRow and
%chunkColumn.
imagesc(newimg)
writecell(resimg2, 'output.csv')

File: MQP_colors.py
import csv
import numpy as np
import pandas as pd
from skimage import io, color
from skimage.transform import resize
from sklearn.cluster import KMeans

# Configuration Variables - feel free to modify these
imageFile = 'geo.png'
n_colors = 16
finalWidth = 160
finalHeight = 160
penColors = ['red','orange','yellow','green','cyan','blue','purple','magenta','black','white','brown','grey']
outputFile = 'output.csv'

# Start - read the image file
original = io.imread(imageFile)

res = resize(original, (finalHeight, finalWidth), order=3) # 0=nearest, 3=BiCubic
oshape = res.shape

# PNG files have an Alpha channel that we do not need. Strip it.
if oshape[2] == 4:
    res = color.rgba2rgb(res)

# Convert to array of 3-color values (r,g,b)
arr = res.reshape((-1, 3))

# Use the KMeans algorithm to find the closest n_colors in the image
kmeans = KMeans(n_clusters=n_colors, random_state=42).fit(arr)
labels = kmeans.predict(arr)
centers = kmeans.cluster_centers_

#reduce the number of colors to what is stated in n_colors
less_colors = centers[labels].reshape(res.shape)

# reduce the array to the unique colors, which should be 16 values
cmap, color_index = np.unique(less_colors.reshape(-1, less_colors.shape[2]), axis=0, return_inverse=True)

# Convert to a Hue Saturation Value (HSV) array
hsvMap = color.rgb2hsv(cmap)

# Break out the HSV values into separate variables for Hue, Saturation, and Value
hue_img = hsvMap[:, 0]
saturation_img = hsvMap[:, 1]
value_img = hsvMap[:, 2]

# Assign color names to the values based on Hue
df = pd.DataFrame({'hue':hue_img})
df['new_colors'] = pd.cut(x=df['hue'],bins=[0, 0.053, 0.1083, 0.1805, 0.4444, 0.5277, 0.7138, 0.8055, 0.8888, 1],
                        labels=['red','orange','yellow','green','cyan','blue','purple','magenta','red'],
                        ordered=False, include_lowest=True)

# Replace Hue color names where Value or Saturation indicate the color is black, white, and brown
df['new_colors'] = np.where((value_img < 0.25), 'black', df.new_colors)
df['new_colors'] = np.where(((saturation_img < 0.05) & (value_img > 0.95)), 'white', df.new_colors)
df['new_colors'] = np.where(((hue_img > 0.53) & (hue_img < 0.1083) & (value_img > 0.25) & (value_img < 0.65)), 'brown', df.new_colors)

# Map each color index value to the corresponding color name (category)
penNameData = [df['new_colors'][v] for v in color_index]
outputPenNames = np.array(penNameData).reshape((-1, finalWidth))

# Assign pen colors to pen numbers
cf = dict([(y, x) for x,y in enumerate(penColors, 1)])
outputPenNumbers = [cf[v] for v in penNameData]

with open(outputFile, 'w', newline='\n') as csvfile:
    csvwriter = csv.writer(csvfile)
    csvwriter.writerows(outputPenNames)